atlp-rwanda/e-commerce-ones-and-zeroes-bn
# [Finishes #187940985] Product Recommendation

### What does this PR do?
Product Recommendation

### Description of task to be completed
This PR introduces a new endpoint for recommending products to users. The primary goal of this feature is to enhance user experience by suggesting products based on specific criteria.

### How should this be manually tested?
this PR can be tested through navigating to the url provided by netlift PR preview: [https://deploy-preview-26--dponesandzereos.netlify.app/](https://github.com/atlp-rwanda/e-commerce-ones-and-zeroes-fn/pull/url)
Then, for the available products, you can view more on a certain product.

### How it's works?

1. First, we retrieve the details of the product using the provided productId.
2. We retrieve all products from the database that are currently available for recommendation.
3. We exclude the current product from the list of available products to avoid recommending the same product that the user is viewing.
4. We identify products that belong to the same collection as the current product. This ensures that the recommendations are relevant in terms of product categories or themes.
5. We look for products with names similar to the current product. This is done by comparing the common part of the product names.
6. We combine the lists of related products and products with similar names, ensuring that there are no duplicates. 
7. We limit the final list of recommended products to a specific number , 5 products) to ensure the recommendations are manageable and useful to the user.


### How should It be tested
1. navigate to the [https://e-commerce-ones-and-zeroes-bn-pd79.onrender.com/docs/](url)
2. Check the availble products within our database
3. Locate the` POST /api/products/recommend` endpoint under the "Products" tag.
4. Prepare a _productId_ of an existing product in your database that you want to base the recommendations on.
5. Try It out
